### ISSUE TITLE
Potential crash in AdsServiceImpl::Save

### ISSUE THREAD

**Issue by iefremov** [brave/brave-browser#14008]
the method allocates `base::ImportantFileWriter` on stack and then calls `WriteNow` which posts to another task runner. The posted tasks uses data from the file writer, which in turn dies immediately, because it is allocated on stack and gets destroyed once the method is finished. This leads to a crash.

Possibly related to https://brave.sp.backtrace.io/p/brave/debug?filters=(_deleted%3D0%2C(ver%2Cregex%2C%228%5B5%7C6%7C7%7C8%7C9%5D.*%22)%2Cptype%3Dbrowser)&fingerprint=641f500fd3cc4c2dd47943d7b526f1d9dd1dca4cfa517256292012ebbfcad042&debug=(c2b275,16,0)
```
[ 00 ] fdatasync
[ 01 ] base::ImportantFileWriter::WriteFileAtomicallyImpl(base::FilePath const&, base::BasicStringPiece<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, base::BasicStringPiece<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, bool)
[ 02 ] base::ImportantFileWriter::WriteScopedStringToFileAtomically(base::FilePath const&, std::__1::unique_ptr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::default_delete<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, base::OnceCallback<void ()>, base::OnceCallback<void (bool)>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
[ 03 ] base::internal::Invoker<base::internal::BindState<void (*)(base::FilePath const&, std::__1::unique_ptr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::default_delete<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, base::OnceCallback<void ()>, base::OnceCallback<void (bool)>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), base::FilePath, std::__1::unique_ptr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::default_delete<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, base::OnceCallback<void ()>, base::OnceCallback<void (bool)>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, void ()>::RunOnce(base::internal::BindStateBase*)

```

**Comment by iefremov** [comment]
The easiest fix would be to just call `WriteFileAtomically` on the task runner, as most of other users of `ImportantFileWriter` do

**Comment by iefremov** [comment]
cc @tmancey

**Comment by iefremov** [comment]
@tmancey I think there is a quick fix, so I'll ask Aleksey to patch it, if you don't mind

**Comment by tmancey** [comment]
@iefremov that would be great if he has time. Otherwise, I can get to this soon

**Comment by btlechowski** [comment]
@iefremov @goodov The issue is marked as `QA/Yes`, but there is no test plan. Please add a test plan or set as `QA/No`

Setting as `QA/Blocked` till resolved.

**Comment by iefremov** [comment]
@btlechowski I think it is ok to leave it alone, adding `QA/No`
